packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -0,0 +1,156 @@
+     * Api to fetch the response as it is for given request.
+     * @param url - url of the request
+     * @param fetchOptions - fetch options for request containing body, headers etc.
+     * @param addInBody - Pass True if caller wants to add epoch in post body.
+     */
+    public async fetchResponse(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<Response> {
+        // Add epoch in fetch request.
+        const request = this.addEpochInRequest(url, fetchOptions, addInBody);
+        try {
+            const response = await fetchHelper(request.url, request.fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    private addEpochInRequest(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean): {url: string, fetchOptions: {[index: string]: any}} {
+        if (this.fluidEpoch !== undefined) {
+            if (addInBody) {
+                // We use multi part form request for post body where we want to use this.
+                // So extract the form boundary to mark the end of form.
+                let body: string = fetchOptions.body;
+                const formBoundary = body.split("\r\n")[0].substring(2);
+                body += `\r\nepoch=${this.fluidEpoch}\r\n`;
+                body += `\r\n--${formBoundary}--`;
+                fetchOptions.body = body;
+            } else {
+                const [mainUrl, queryString] = url.split("?");
+                const searchParams = new URLSearchParams(queryString);
+                searchParams.append("epoch", this.fluidEpoch);
+                const urlWithEpoch = `${mainUrl}?${searchParams.toString()}`;
+                if (urlWithEpoch.length > 2048) {
+                    // Add in headers if the length becomes greater than 2048
+                    // as ODSP has limitation for queries of length more that 2048.
+                    fetchOptions.headers = {
+                        ...fetchOptions.headers,
+                        "x-fluid-epoch": this.fluidEpoch,
+                    };
+                } else {
+                    return {
+                        url: urlWithEpoch,
+                        fetchOptions,
+                    };
+                }
+            }
+        }
+        return { url, fetchOptions };
+    }
+
+    private validateEpochFromResponse(epochFromResponse: string | undefined | null) {
+        // If epoch is undefined, then don't compare it because initially for createNew or TreesLatest
+        // initializes this value. Sometimes response does not contain epoch as it is still in
+        // implementation phase at server side. In that case also, don't compare it with our epoch value.
+        assert(this.fluidEpoch === undefined
```

Comment:
This assert will not trigger check in checkForEpochError() and thus cache would not be cleared

Reply:
Also when can epochFromResponse be null? If it can't be null, I'd remove it. If you want to be safe, you can use falsy check for all of these, as "" is also not valid epoch

Reply:
Added throw error with 409 status code in this api so that  "checkForEpochError" can catch it.

Reply:
Yes, I wanted to be safe. Moved to falsy checks.

Reply:
Also, this is just extra layer of security because server will already throw the error if epoch will mismatch.